### PR TITLE
DAOS-7821 build: Add a gofmt githook

### PR DESCRIPTION
**Pull request by mjmac**
Require that any changes to .go files be run through
gofmt before creating a PR.

Required-githooks: true

Signed-off-by: Michael MacDonald <mjmac.macdonald@intel.com>

### CODE REVIEW

File: utils/githooks/pre-commit.d/40-gofmt.sh [hunk 1]
```diff
@@ -0,0 +1,47 @@
+#!/bin/sh
+
+# Runs pylint for the DAOS project as a commit hook.
```
Comment (ashleypittman) [inline utils/githooks/pre-commit.d/40-gofmt.sh:3]:
```suggestion
# Runs gofmt for the DAOS project as a commit hook.
```

File: utils/githooks/pre-commit.d/40-gofmt.sh [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+if [ $TARGET = "HEAD" ]
+then
+        echo "Checking against HEAD"
+        output=$(git diff HEAD^ --name-only | grep -e '.go$' | xargs gofmt -d)
+else
+        echo "Checking against branch ${TARGET}"
+        output=$(git diff $TARGET... --name-only | grep -e '.go$' | xargs gofmt -d)
+fi
+
+if [ -n "$output" ]; then
```
Comment (ashleypittman) [inline utils/githooks/pre-commit.d/40-gofmt.sh:39]: We have other commit-hooks (including but not limited to the spelling one) that just auto-applies the fixes, although this is probably good enough.  The problem with auto-applying is you need to run "git add" afterwards in the commit hook for modified files and there's no easy way to know which they are.

File: utils/githooks/pre-commit.d/40-gofmt.sh [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+        echo "ERROR: Your code hasn't been run through gofmt!"
+        echo "Please configure your editor to run gofmt on save."
+        echo "Alternatively, at a minimum, run the following command:"
+        echo -n "find src/control -name '*.go' -and -not -path '*vendor*'"
```
[the remark this answers is not in the document]
Comment (jolivier23) [reply utils/githooks/pre-commit.d/40-gofmt.sh:43]: maybe just make it /bin/bash?

File: utils/githooks/pre-commit.d/40-gofmt.sh (hunk 2 above)
Comment (mjmac) [reply utils/githooks/pre-commit.d/40-gofmt.sh:39]: Yeah, most people touching .go files will have their editor set up to run gofmt on save anyhow.

File: utils/githooks/pre-commit.d/40-gofmt.sh [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+output=
+if [ $TARGET = "HEAD" ]
+then
+        echo "Checking against HEAD"
+        output=$(git diff HEAD^ --name-only | grep -e '.go$' | xargs gofmt -d)
+else
+        echo "Checking against branch ${TARGET}"
+        output=$(git diff $TARGET... --name-only | grep -e '.go$' | xargs gofmt -d)
+fi
```
Comment (brianjmurrell) [inline utils/githooks/pre-commit.d/40-gofmt.sh:37]: Not saying necessarily for this PR, but this code is begging to be DRYed out.  We have it in a number of hooks now.

File: utils/githooks/pre-commit.d/40-gofmt.sh (hunk 2 above)
Comment (ashleypittman) [reply utils/githooks/pre-commit.d/40-gofmt.sh:39]: > > The problem with auto-applying is you need to run "git add" afterwards in the commit hook for modified files and there's no easy way to know which they are
> 
> Doesn't the copyright githook do exactly that?

https://github.com/daos-stack/daos/blob/master/utils/githooks/pre-commit.d/10-update-copyright#L63 Yes, although in that case it uses grep to know which files to run it on.  My point was we might not want to run it for every go file.

**Review by ashleypittman — APPROVED** [review]

**Review by jolivier23 — APPROVED** [review]

**Pull request merged**